tests/baselines/reference/awaitUsingDeclarations.2(target=es2015).js

```diff
@@ -0,0 +1,80 @@
+}
+
+export {};
+
+//// [awaitUsingDeclarations.2.js]
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
+    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
+    return new (P || (P = Promise))(function (resolve, reject) {
+        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
+        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
+        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
+        step((generator = generator.apply(thisArg, _arguments || [])).next());
+    });
+};
+var __addDisposableResource = (this && this.__addDisposableResource) || function (env, value, async) {
+    if (value !== null && value !== void 0) {
+        if (typeof value !== "object") throw new TypeError("Object expected.");
+        var dispose;
+        if (async) {
+            if (!Symbol.asyncDispose) throw new TypeError("Symbol.asyncDispose is not defined.");
+            dispose = value[Symbol.asyncDispose];
+        }
+        if (dispose === void 0) {
+            if (!Symbol.dispose) throw new TypeError("Symbol.dispose is not defined.");
+            dispose = value[Symbol.dispose];
+        }
+        if (typeof dispose !== "function") throw new TypeError("Object not disposable.");
+        env.stack.push({ value: value, dispose: dispose, async: async });
+    }
+    else if (async) {
+        env.stack.push({ async: true });
+    }
+    return value;
+};
+var __disposeResources = (this && this.__disposeResources) || function (env) {
+    function fail(e) {
+        env.error = env.hasError ? new SuppressedError(e, env.error) : e;
+        env.hasError = true;
+    }
+    function next() {
+        while (env.stack.length) {
+            var rec = env.stack.pop();
+            try {
+                var result = rec.dispose && rec.dispose.call(rec.value);
+                if (rec.async) return Promise.resolve(result).then(next, function(e) { fail(e); return next(); });
+            }
+            catch (e) {
+                fail(e);
+            }
+        }
+        if (env.hasError) throw env.error;
+    }
+    return next();
+};
+{
+    const env_1 = { stack: [], error: void 0, hasError: false };
+    try {
+        const d1 = __addDisposableResource(env_1, { [Symbol.asyncDispose]() {
+                return __awaiter(this, void 0, void 0, function* () { });
+            } }, true), d2 = __addDisposableResource(env_1, { [Symbol.asyncDispose]() {
+                return __awaiter(this, void 0, void 0, function* () { });
+            } }, true);
+    }
+    catch (e_1) {
+        env_1.error = e_1;
+        env_1.hasError = true;
+    }
+    finally {
+        const result_1 = __disposeResources(env_1);
+        if (result_1)
+            await result_1;

```

Comment:
What this `if` needed for? In Babel we `await` unconditionally, and I'm wondering if it's a bug. Or is it somehow related to https://github.com/tc39/proposal-async-explicit-resource-management/issues/17?

Btw, this `await` technically adds one extra tick that is not in the spec. If there is 1 async resource, the total number of `await`s is 2 instead of 1 (once inside the helper, and once here). We decided that we are ok with this spec deviation in Babel, but I just want to make sure that you noticed it.

Reply:
Per the specified semantics, an `await using` declaration only causes an implicit `await` if it is evaluated:

```js
switch (x) {
  case 0:
    await using y = getAsyncResource();
    // falls through
  case 1:
    using z = getSyncResource();
    break;
}
```

Despite both `y` and `z` existing in the same block scope, there is only an implicit `await` if `x === 0`. When `x === 1`, the `await using` is not evaluated, thus should not impose an implicit await at the end of the block.

Here is another example:

```js
outer: {
  using z = getSyncResource();
  if (x === 1) break outer;
  await using y = getAsyncResource();
}
```

Again, if `x === 1` then the `await using` is not evaluated, and thus does not impose an implicit `await`.

However, there is always an implicit `await` if an `await using` *is* evaluated, even if its initializer is `null` or `undefined`:

```js
{
  using z = null;
  await using y = null;
} // implicit `await` even if there is nothing to dispose
```

I have yet to finish the test262 tests that would be used to validate this, though that work is currently underway.

Reply:
> Btw, this `await` technically adds one extra tick that is not in the spec. If there is 1 async resource, the total number of `await`s is 2 instead of 1 (once inside the helper, and once here). We decided that we are ok with this spec deviation in Babel, but I just want to make sure that you noticed it.

I think this is mostly unavoidable in downlevel emit without a significant amount of inlining.

Reply:
Oh, thank you!